collect_app/src/test/java/org/odk/collect/android/external/FormUriActivityTest.kt

```diff
@@ -0,0 +1,158 @@
+    private val context = ApplicationProvider.getApplicationContext<Application>()
+    private val firstProject = Project.DEMO_PROJECT
+    private val secondProject = Project.Saved("123", "Second project", "S", "#cccccc")
+
+    @Before
+    fun setup() {
+        projectsRepository = InMemProjectsRepository()
+        currentProjectProvider = mock()
+
+        projectsRepository.save(firstProject)
+        projectsRepository.save(secondProject)
+        whenever(currentProjectProvider.getCurrentProject()).thenReturn(firstProject)
+
+        CollectHelpers.overrideAppDependencyModule(object : AppDependencyModule() {
+            override fun providesProjectsRepository(
+                uuidGenerator: UUIDGenerator?,
+                gson: Gson?,
+                settingsProvider: SettingsProvider?
+            ): ProjectsRepository {
+                return projectsRepository
+            }
+
+            override fun providesCurrentProjectProvider(
+                settingsProvider: SettingsProvider?,
+                projectsRepository: ProjectsRepository?
+            ): CurrentProjectProvider {
+                return currentProjectProvider
+            }
+        })
+    }
+
+    @Test
+    fun `When there are no projects then display alert dialog`() {
+        projectsRepository.deleteAll()
+
+        val scenario = ActivityScenario.launch(FormUriActivity::class.java)
+        onView(withText(R.string.no_projects_detected)).inRoot(isDialog()).check(matches(isDisplayed()))
+        onView(withId(android.R.id.button1)).perform(click())
+
+        assertThat(scenario.result.resultCode, `is`(Activity.RESULT_CANCELED))
+    }
+
+    @Test
+    fun `When there is project id specified in uri and it does not match current project id then display alert dialog`() {
+        val scenario = ActivityScenario.launch<FormUriActivity>(getIntent(secondProject.uuid))
+
+        onView(withText(R.string.wrong_project_selected_for_form)).inRoot(isDialog()).check(matches(isDisplayed()))
+        onView(withId(android.R.id.button1)).perform(click())
+
+        assertThat(scenario.result.resultCode, `is`(Activity.RESULT_CANCELED))
+    }
+
+    @Test
+    fun `When there is project id specified in uri and it matches current project id then start form filling`() {
+        Intents.init()
+
+        ActivityScenario.launch<FormUriActivity>(getIntent(firstProject.uuid))
+
+        Intents.intended(hasComponent(FormEntryActivity::class.java.name))
+
+        Intents.release()
+    }
+
+    @Test
+    fun `When there is no project id specified in uri and first available project id does not match current project id then display alert dialog`() {
+        whenever(currentProjectProvider.getCurrentProject()).thenReturn(secondProject)
+
+        val scenario = ActivityScenario.launch<FormUriActivity>(getIntent())
+
+        onView(withText(R.string.wrong_project_selected_for_form)).inRoot(isDialog()).check(matches(isDisplayed()))
```

Comment:
I did not know that we could use matchers on dialogs in Robolectric. That's definitely useful.